File: backend/app/pipeline/chunk_embed.py
import hashlib

from openai import OpenAI

from app.core.config import Settings
from app.db.database import SessionLocal
from app.db.repository import list_documents, save_documents
from app.rag.faiss_store import FaissStore


def _hash_text(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def save_raw_documents(settings: Settings, documents: list[dict]) -> int:
    with SessionLocal() as db:
        return save_documents(db, documents)


def load_raw_documents(settings: Settings) -> list[dict]:
    with SessionLocal() as db:
        return list_documents(db)


def split_text(text: str, chunk_size: int, chunk_overlap: int) -> list[str]:
    clean = " ".join(text.split())
    if len(clean) <= chunk_size:
        return [clean]

    chunks: list[str] = []
    start = 0
    step = max(chunk_size - chunk_overlap, 1)

    while start < len(clean):
        end = min(start + chunk_size, len(clean))
        chunks.append(clean[start:end])
        if end == len(clean):
            break
        start += step

    return chunks


def build_chunk_records(documents: list[dict], settings: Settings) -> list[dict]:
    chunks: list[dict] = []

    for document in documents:
        split_chunks = split_text(
            text=document.get("text", ""),
            chunk_size=settings.chunk_size,
            chunk_overlap=settings.chunk_overlap,
        )

        for idx, chunk_text in enumerate(split_chunks):
            if not chunk_text.strip():
                continue
            chunks.append(
                {
                    "chunk_id": _hash_text(document.get("source_url", "") + str(idx) + chunk_text),
                    "source_url": document.get("source_url", ""),
                    "title": document.get("title", "Untitled"),
                    "text": chunk_text,
                }
            )

    return chunks


def embed_chunks(chunks: list[dict], settings: Settings) -> list[list[float]]:
    if not settings.active_api_key:
        raise ValueError("Missing active LLM API key for selected provider")

    client = OpenAI(api_key=settings.active_api_key, base_url=settings.active_base_url)
    embeddings: list[list[float]] = []

    batch_size = 64
    for start in range(0, len(chunks), batch_size):
        batch = chunks[start : start + batch_size]
        request_kwargs = {
            "model": settings.llm_embedding_model,
            "input": [item["text"] for item in batch],
        }
        if settings.llm_embedding_document_input_type.strip():
            request_kwargs["input_type"] = settings.llm_embedding_document_input_type.strip()

        response = client.embeddings.create(**request_kwargs)
        embeddings.extend([row.embedding for row in response.data])

    return embeddings


def rebuild_faiss_index(settings: Settings) -> dict:
    documents = load_raw_documents(settings)
    if not documents:
        return {"documents": 0, "chunks": 0}

    chunks = build_chunk_records(documents, settings)
    if not chunks:
        return {"documents": len(documents), "chunks": 0}

    vectors = embed_chunks(chunks, settings)

    store = FaissStore(
        index_path=settings.index_dir / "index.faiss",
        metadata_path=settings.index_dir / "metadata.jsonl",
    )
    indexed = store.save(embeddings=vectors, metadata=chunks)

    return {"documents": len(documents), "chunks": indexed}

File: backend/src/pipeline/chunk_embed.ts
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";
import OpenAI from "openai";
import { getSettings } from "../core/config";
import {
  saveChunkEmbeddingsBatch,
  deleteOldChunkEmbeddings,
} from "../db/repository";
import logger from "../core/logger";

export interface RawDocument {
  source_url: string;
  title: string;
  text: string;
}

export interface ChunkRecord {
  chunk_id: string;
  source_url: string;
  title: string;
  text: string;
}

export interface RebuildResult {
  documents: number;
  chunks: number;
}

function hashText(text: string): string {
  return crypto.createHash("sha256").update(text, "utf-8").digest("hex");
}

export async function saveRawDocuments(
  documents: RawDocument[]
): Promise<number> {
  const settings = getSettings();

  // Ensure directory exists
  if (!fs.existsSync(settings.dataDir)) {
    fs.mkdirSync(settings.dataDir, { recursive: true });
  }

  const filePath = settings.rawDocsFile;

  // Append documents to JSONL file
  for (const doc of documents) {
    fs.appendFileSync(filePath, JSON.stringify(doc) + "\n");
  }

  return documents.length;
}

export async function loadRawDocuments(): Promise<RawDocument[]> {
  const settings = getSettings();
  const filePath = settings.rawDocsFile;

  if (!fs.existsSync(filePath)) {
    return [];
  }

  const content = fs.readFileSync(filePath, "utf-8");
  const documents: RawDocument[] = [];

  for (const line of content.split("\n")) {
    if (line.trim()) {
      try {
        documents.push(JSON.parse(line));
      } catch (e) {
        logger.warn("Failed to parse document line:", line);
      }
    }
  }

  return documents;
}

export function splitText(
  text: string,
  chunkSize: number,
  chunkOverlap: number
): string[] {
  const clean = text.split(/\s+/).join(" ");

  if (clean.length <= chunkSize) {
    return [clean];
  }

  const chunks: string[] = [];
  const step = Math.max(chunkSize - chunkOverlap, 1);

  let start = 0;
  while (start < clean.length) {
    const end = Math.min(start + chunkSize, clean.length);
    chunks.push(clean.substring(start, end));

    if (end === clean.length) {
      break;
    }

    start += step;
  }

  return chunks;
}

export function buildChunkRecords(
  documents: RawDocument[],
  settings: any
): ChunkRecord[] {
  const chunks: ChunkRecord[] = [];

  for (const document of documents) {
    const splitChunks = splitText(
      document.text,
      settings.chunkSize,
      settings.chunkOverlap
    );

    for (let idx = 0; idx < splitChunks.length; idx++) {
      const chunkText = splitChunks[idx];
      if (!chunkText.trim()) {
        continue;
      }

      chunks.push({
        chunk_id: hashText(document.source_url + String(idx) + chunkText),
        source_url: document.source_url,
        title: document.title || "Untitled",
        text: chunkText,
      });
    }
  }

  return chunks;
}

export async function embedChunks(
  chunks: ChunkRecord[]
): Promise<number[][]> {
  const settings = getSettings();

  if (!settings.getActiveApiKey()) {
    throw new Error("Missing active LLM API key for selected provider");
  }

  const client = new OpenAI({
    apiKey: settings.getActiveApiKey(),
    baseURL: settings.getActiveBaseUrl(),
  });

  const embeddings: number[][] = [];
  const batchSize = 64;

  for (let start = 0; start < chunks.length; start += batchSize) {
    const batch = chunks.slice(start, start + batchSize);
    const requestParams: any = {
      model: settings.llmEmbeddingModel,
      input: batch.map((item) => item.text),
    };

    if (settings.llmEmbeddingDocumentInputType) {
      requestParams.input_type =
        settings.llmEmbeddingDocumentInputType;
    }

    try {
      const response = await client.embeddings.create(requestParams);
      for (const row of response.data) {
        embeddings.push(row.embedding);
      }
    } catch (error) {
      logger.error("Embedding batch failed:", error);
      throw error;
    }
  }

  return embeddings;
}

export async function rebuildFaissIndex(): Promise<RebuildResult> {
  const settings = getSettings();

  try {
    const documents = await loadRawDocuments();
    if (documents.length === 0) {
      return { documents: 0, chunks: 0 };
    }

    const chunks = buildChunkRecords(documents, settings);
    if (chunks.length === 0) {
      return { documents: documents.length, chunks: 0 };
    }

    logger.info(`Embedding ${chunks.length} chunks...`);
    const vectors = await embedChunks(chunks);

    // Clear old embeddings
    await deleteOldChunkEmbeddings();

    // Save new embeddings
    const chunkEmbeddings = chunks.map((chunk, idx) => ({
      chunkId: chunk.chunk_id,
      sourceUrl: chunk.source_url,
      title: chunk.title,
      text: chunk.text,
      embedding: vectors[idx],
      vectorNorm: 0, // Will be computed on retrieval
    }));

    await saveChunkEmbeddingsBatch(chunkEmbeddings);

    return { documents: documents.length, chunks: chunks.length };
  } catch (error) {
    logger.error("Failed to rebuild index:", error);
    throw error;
  }
}
